google/cloud/ndb/_datastore_api.py

```diff
@@ -316,8 +316,7 @@ def get_read_options(options, default_read_consistency=None):
         ValueError: When ``read_consistency`` is set to ``EVENTUAL`` and there
             is a transaction.
     """
-    transaction = _get_transaction(options)
-
+    transaction = options.transaction
```

Comment:
So, what was going on was the transaction would be set or not for the batch when `_datastore_api.lookup` was called, but the call to `_get_transaction` would potentially fish the transaction out of the current context when the batch was actually run. The problem is that the call to `lookup` would be made in the context of a `non_transactional` and would correctly set the transaction to `None`, but then because the call to actually process the batch was delayed, it could happen outside of the `non_transactional` context, even though it is processing `lookup` calls made in the `non_transactional` context. The solution here is to just use the transaction set for the batch without referring to the current context. I'm pretty sure the call to `_get_transaction` is a fossil from a time before the transactions feature had been completely figured out. I have verified that in all cases the current transaction is set in `options` when the `lookup` call is made, so there really is no need to refer to the context.